src/sparse/impl/KokkosSparse_spmv_impl.hpp

```diff
@@ -502,380 +502,449 @@ struct SPMV_MV_Transpose_Functor {
   }
 };
 
+
+
+
 template<class AMatrix,
          class XVector,
          class YVector,
          int doalpha,
          int dobeta,
          bool conjugate>
-struct SPMV_MV_LayoutLeft_Functor {
-  typedef typename AMatrix::execution_space            execution_space;
-  typedef typename AMatrix::non_const_ordinal_type               ordinal_type;
-  typedef typename AMatrix::non_const_value_type       A_value_type;
-  typedef typename YVector::non_const_value_type       y_value_type;
-  typedef typename Kokkos::TeamPolicy<execution_space> team_policy;
-  typedef typename team_policy::member_type            team_member;
-  typedef typename YVector::non_const_value_type       coefficient_type;
+struct SPMV_MV_Functor {
+  typedef typename AMatrix::execution_space              execution_space;
+  typedef typename AMatrix::non_const_ordinal_type       ordinal_type;
+  typedef typename AMatrix::non_const_value_type         value_type;
+  typedef typename Kokkos::TeamPolicy<execution_space>   team_policy;
+  typedef typename team_policy::member_type              team_member;
+  typedef Kokkos::Details::ArithTraits<value_type>       ATV;
 
-  const coefficient_type alpha;
-  AMatrix m_A;
+  const value_type alpha;
+  AMatrix  m_A;
   XVector m_x;
-  const coefficient_type beta;
+  const value_type beta;
   YVector m_y;
-  //! The number of columns in the input and output MultiVectors.
-  ordinal_type n;
-  ordinal_type rows_per_thread;
-
-  SPMV_MV_LayoutLeft_Functor (const coefficient_type& alpha_,
-                              const AMatrix& m_A_,
-                              const XVector& m_x_,
-                              const coefficient_type& beta_,
-                              const YVector& m_y_,
-                              const ordinal_type rows_per_thread_) :
-    alpha (alpha_),
-    m_A (m_A_), m_x (m_x_), beta (beta_), m_y (m_y_), n (m_x_.extent(1)),
-    rows_per_thread (rows_per_thread_)
-  {}
 
-  template<int UNROLL>
-  KOKKOS_INLINE_FUNCTION void
-  strip_mine (const team_member& dev, const ordinal_type& iRow, const ordinal_type& kk) const
-  {
-    y_value_type sum[UNROLL];
+  const ordinal_type rows_per_team;
+  const ordinal_type numVecs;
+  const ordinal_type vecRemainder;
+
+  struct value16 {
```

Comment:
Does kokkos-kernels already have a SIMD type?

Reply:
True, I am waiting for more feedback on the simd type before going ahead and using them in this kernel